filebeat/inputsource/unix/server.go

```diff
@@ -84,8 +125,77 @@ func (s *Server) createServer() (net.Listener, error) {
 	return l, nil
 }
 
-func (s *Server) cleanupStaleSocket() error {
-	path := s.config.Path
+func (s *datagramServer) createConn() (net.Conn, error) {
+	if err := cleanupStaleSocket(s.config.Path); err != nil {
+		return nil, err
+	}
+
+	addr, err := net.ResolveUnixAddr("unixgram", s.config.Path)
+	if err != nil {
+		return nil, err
+	}
+	conn, err := net.ListenUnixgram("unixgram", addr)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := setSocketOwnership(s.config.Path, s.config.Group); err != nil {
+		return nil, err
+	}
+
+	if err := setSocketMode(s.config.Path, s.config.Mode); err != nil {
+		return nil, err
+	}
+	return conn, nil
+}
+
+func (s *datagramServer) Run(ctx context.Context) error {
+	conn, err := s.createConn()
+	if err != nil {
+		return err
+	}
+
+	for ctx.Err() == nil {
+		err := s.connHandler(ctx, conn)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *datagramServer) Start() error {
+	conn, err := s.createConn()
+	if err != nil {
+		return err
+	}
+
+	s.tg.Go(func(ctx unison.Canceler) error {
+		connCtx, connCancel := ctxtool.WithFunc(ctxtool.FromCanceller(ctx), func() {
+			conn.Close()
+		})
+		defer connCancel()
+
+		for connCtx.Err() == nil {
+			err := s.connHandler(ctxtool.FromCanceller(connCtx), conn)
```

Comment:
Not sure where connHandler is defined. But is this one special for the datagram socket?

The `UnixConn` type implements methods for streaming and methods for datagram. Unfortunately the type implements the [`PacketConn`](https://pkg.go.dev/net/#PacketConn) (datagram and raw packets), and [`Conn`](https://pkg.go.dev/net/#Conn)(network streaming) interfaces. In case of datagram we want the `PacketConn` interface, similar to the UDP inputsource.

The `common.ConnectionHandler` interface looks like the wrong type for our use-case. The handlers for TCP have some kind of event splitting support. The UDP one expects one packets == one event.

I'm sure the `read` syscall will work for both, but the semantics will be different.

Reply:
Haha, yes. It accidentally worked in my tests. :D